crates/bevy_ecs/src/system/query/extensions.rs

```diff
@@ -0,0 +1,63 @@
+use std::fmt::Debug;
+
+use thiserror::Error;
+
+use crate::{Fetch, Query, QueryFilter, ReadOnlyFetch, WorldQuery};
+
+impl<'a, Q: WorldQuery, F: QueryFilter> Query<'a, Q, F> {
+    /// Takes exactly one result from the query. If there are no results, or more than 1 result, this will return an error instead.
+    pub fn get_unique(&self) -> Result<<Q::Fetch as Fetch<'_>>::Item, OnlyQueryError<'_, Q>>
+    where
+        Q::Fetch: ReadOnlyFetch,
+    {
+        let mut query = self.iter();
+        let first = query.next();
+        let extra_count = query.count();
```

Comment:
I'm not sure `query.count()` is the right tool here. In the "happy path" case where there is only one entity it will have a linear cost. But if someone wants to call get_unique() as a check on a potentially large query, then suddenly its O(n) (because count() calls iter.next() until it returns None).

I don't think returning the actual count in the error is worth the runtime cost of the check. I think it makes more sense to just call iter.next() a second time to check if it is None. Users that need the full count can always detect the error and call query.iter().count().